enhancements/testing/improved-platform-tests.md

```diff
@@ -0,0 +1,412 @@
+   environment variable, for example).
+
+### Risks and Mitigations
+
+- The current policy when the [origin](https://github.com/openshift/origin/) pull
+  request bringing the updated k8s version breaks the CI is to revert PR. This
+  policy will be harder to execute with this change because reverting the o/k
+  rebase PR can be problematic and was never tried before.
+
+There should be fewer reasons to do any reverting, because with this new mechanism
+we should get the appropriate signal during extensive payload testing required
+to land every k8s bump PR.
+
+- Some of the work that has been done to build the partner certification tool will
+  need to be updated to pull additional images with separate test binaries.
+
+The `openshift-tests` binary will contain the necessary logic how to retrieve those
+additional tests binaries from the release payload, so even in the offline installation
+as long as the release payload is reachable, the tests should continue working.
+
+- Not available release image can prevent from running the tests.
+
+The proposed mechanism will contain a fallback path, which will ensure that in
+cases where the release payload is not reachable we will be able to run all the
+tests bundled within the `openshift-tests` binary. This includes the basic
+OpenShift and Kubernetes conformance tests.
+
+- Slowing down the local development cycle due to pulling the external test binaries.
+
+The initial proposal is to use an environment variable `OPENSHIFT_SKIP_EXTERNAL_TESTS`,
+which bypasses the mechanism and always uses the tests bundled in `openshift-tests`
+binary.
+At a later point in time, it's reasonable to re-use test binaries downloaded
+in prior executions, through exposing an environment variable,
+`OPENSHIFT_EXTERNAL_TESTS_DIRECTORY` for example.
+
+- In some cases, after landing newer version of kubernetes it is possible that
+  we'll be having issues with tests relying on newer kubelets, which is usually
+  shipped with updated base image with some delay.
+
+This problem will not present during PR testing, because then we always build
+the base image from scratch to ensure all components are using newest versions
+for tests. For those rare cases where tests will rely on newer functionality in
+kubelet we can either temporarily disable the test before landing the PR, or
+sync with ART to ensure the updated kubelet is shipped as quickly as possible.
+
+- Upgrade tests, unlike other tests are executed within single test binary
+  making it harder to migrate to use external test binaries.
+
+The current upgrade test suite runs only a handful of tests which don't change
+that often. For that reason, we've initially decided to rely only on the vendored
+code for the upgrade tests.
+
+### Drawbacks
+
+This proposal might lead to unnecessary proliferation of external tests binaries,
+which might cause problems when the release image is not available, or for local
+development.
```

Comment:
hard to imagine a situation where the release image is not available to someone who's trying to run tests

Reply:
What about microshift?

Reply:
i'll have to defer to @dhellmann on how they manage this today (where they get the openshift-tests binary from, how they distribute it to customers) on the implications there, but in my mind it's another argument for why we're better off having a single test image that contains all the test binaries.  Microshift can then consume/reference that image.

Reply:
There is no release payload for MicroShift. It looks like the CI job that uses openshift-tests pulls it from an image built as part of the CI system? https://github.com/openshift/release/blob/master/ci-operator/config/openshift/microshift/openshift-microshift-main.yaml#L47

@pacevedom, @pmtk , or @copejon  can any of you explain where that binary comes from?

We do not ship the openshift-tests binary as part of MicroShift.

Reply:
Following the breadcrumbs it's: inputs.test-bin which is 
```
  test-bin:
    name: "4.14"
    namespace: ocp
    tag: tests
```
Which is imported: `Tagging ocp/4.14:tests into pipeline:test-bin`.
I think it's [this image config](https://github.com/openshift/release/blob/master/ci-operator/config/openshift/origin/openshift-origin-master.yaml#L26-L32)

tl;dr: we just take it from a promoted image

Reply:
So we may be OK in CI. How would someone run the tests by hand, like if they were debugging a problem?

Reply:
Not sure, I'd build `openshift-tests` from source. Maybe @pacevedom has better idea

Reply:
Based on this proposal from @soltysh , building from source wouldn't give us the tests that had been moved out of the repo.

@soltysh or @bparees, do you have something in mind for the workflow where someone is trying to test outside of our CI system and has no release payload?

Reply:
>  the workflow where someone is trying to test outside of our CI system and has no release payload?

if they aren't having to modify the tests themselves, they could still use the test binaries from a recent payload image(since @soltysh's current proposal is that there would be multiple test images, one per binary, i defer to him on exactly how a consumer/client identifies all the images needed, pulls them, and extracts all the binaries, but that's something that CI will need to be doing too and i expect it will be relatively straightforward and handled by the openshift-tests binary itself)

if they are modifying the tests locally, then i'd expect them to be able to:
1) build/get a copy of the openshift-tests wrapper binary
2) invoke it while pointing it at their local externalized test binary for their component

(it would only run the tests that are built into openshift-tests plus the tests in their binary, but usually when people are doing this they are already narrowing the set of tests down to a few specific ones anyway).

Do those two options cover your use cases?

Reply:
Number 2 is covered in risks sections, where I'm talking about possibility to use locally available binaries instead of pulling them from release. It's not implemented, but it's rather simple. I'll add it to future work section.

Reply:
The only downside is you need to know which binaries you care about, but in case of microshift it's rather simple b/c you'll most likely care about k8s mostly.

Reply:
Based on [Fabio's comment](https://github.com/openshift/enhancements/pull/1291#discussion_r1246863838) we'll stick with in-binary k8s tests for now, and we'll slowly work our way out in the future, remembering about cases like CSI and Microshift.